Internal error on mutation

This mutation produces an internal error (Note this is a target shared by more than one observation)

```graphql
mutation($input: CloneTargetInput!) {
  cloneTarget(input: $input) {
    newTarget {
      id
    }
  }
}
```

```json
{
    "input": {
        "targetId": "t-105",
        "REPLACE_IN": [
            "o-10a"
        ]
    }
}
```

Produces this response:
```json
{
  "errors": [
    {
      "message": "An internal error of type PostgresErrorException occurred."
    }
  ]
}
```

On the server I see this message as well:

```
service [ERROR] lucuma-odb - Error computing GraphQL response. 
service skunk.exception.PostgresErrorException: 
service 🔥  
service 🔥  Postgres ERROR 23503 raised in ri_ReportViolation (ri_triggers.c:2609)
service 🔥  
service 🔥    Problem: Update or delete on table "t_asterism_target" violates foreign key
service 🔥             constraint
service 🔥             "t_itc_result_c_program_id_c_observation_id_c_target_id_fkey" on
service 🔥             table "t_itc_result".
service 🔥     Detail: Key (c_program_id, c_observation_id, c_target_id)=(p-101, o-10a,
service 🔥             t-105) is still referenced from table "t_itc_result".
service 🔥  
service 🔥  The statement under consideration was defined
service 🔥    at /Users/carlos.quiroz/code/noirlab/lucuma-odb/modules/service/src/main/scala/lucuma/odb/service/TargetService.scala:453
service 🔥  
service 🔥    UPDATE t_asterism_target
service 🔥    SET    c_target_id = $1
service 🔥    WHERE  c_target_id = $2
service 🔥    AND    c_observation_id IN ($3)
service 🔥  
service 🔥  and the arguments were provided
service 🔥    at /Users/carlos.quiroz/code/noirlab/lucuma-odb/modules/service/src/main/scala/lucuma/odb/service/TargetService.scala:169
service 🔥  
service 🔥    $1 varchar    t-108
service 🔥    $2 varchar    t-105
service 🔥    $3 varchar    o-10a
service 🔥  
service 🔥  If this is an error you wish to trap and handle in your application, you can do
service 🔥  so with a SqlState extractor. For example:
service 🔥  
service 🔥    doSomething.recoverWith { case SqlState.ForeignKeyViolation(ex) =>  ...}
service 🔥  
service 
service skunk.exception.PostgresErrorException: Update or delete on table "t_asterism_target" violates foreign key constraint "t_itc_result_c_program_id_c_observation_id_c_target_id_fkey" on table "t_itc_result".
```

--- a/modules/service/src/main/scala/lucuma/odb/service/TargetService.scala
+++ b/modules/service/src/main/scala/lucuma/odb/service/TargetService.scala
@@ -165,8 +165,10 @@ object TargetService {
 
         def replaceIn(tid: Target.Id): F[Unit] =
           input.REPLACE_IN.traverse_ { which =>
-            val stmt = Statements.replaceTargetIn(which, input.targetId, tid)
-            session.prepareR(stmt.fragment.command).use(_.execute(stmt.argument))
+            val s1 = Statements.dropItcCache(which)
+            val s2 = Statements.replaceTargetIn(which, input.targetId, tid)
+            session.prepareR(s1.fragment.command).use(_.execute(s1.argument)) >>
+            session.prepareR(s2.fragment.command).use(_.execute(s2.argument))
           }
 
         pid.flatMap {
@@ -448,6 +450,12 @@ object TargetService {
       void"""
         RETURNING c_target_id
       """
+
+    def dropItcCache(which: NonEmptyList[Observation.Id]): AppliedFragment =
+      sql"""
+      DELETE FROM t_itc_result
+      WHERE  c_observation_id IN (${observation_id.list(which.length)})
+      """.apply(which.toList)
 
     def replaceTargetIn(which: NonEmptyList[Observation.Id], from: Target.Id, to: Target.Id): AppliedFragment =
       sql"""

--- a/modules/service/src/test/scala/lucuma/odb/graphql/issue/github/687.scala
+++ b/modules/service/src/test/scala/lucuma/odb/graphql/issue/github/687.scala
@@ -1,0 +1,78 @@
+// Copyright (c) 2016-2023 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package lucuma.odb.graphql
+package issue.github
+
+import io.circe.literal.*
+import io.circe.syntax.*
+import lucuma.core.model.Target
+import lucuma.odb.graphql.query.ObservingModeSetupOperations
+
+// https://github.com/gemini-hlsw/lucuma-odb/issues/687
+class GitHub_687 extends OdbSuite with ObservingModeSetupOperations {
+  val pi = TestUsers.Standard.pi(nextId, nextId)
+  val validUsers = List(pi)
+
+  test("clone target with replace_in, where obs has cached ITC results") {
+    for
+      pid <- createProgramAs(pi)
+      tid <- createTargetWithProfileAs(pi, pid)
+      o1  <- createGmosNorthLongSlitObservationAs(pi, pid, List(tid))
+      _   <- expect(
+        user = pi,
+        query = s"""
+        query {
+          observation(observationId: ${o1.asJson}) {
+            itc {
+              science {
+                index
+              }
+            }
+          }
+        }
+        """,
+        expected = Right(
+          json"""
+            {
+              "observation" : {
+                "itc" : {
+                  "science" : {
+                    "index" : 0
+                  }
+                }
+              }
+            }
+          """
+        )
+      )
+      _   <- expect(
+        user = pi,
+        query = s"""
+          mutation {
+            cloneTarget(input: {
+              targetId: ${tid.asJson}
+              REPLACE_IN: [${o1.asJson}]
+            }) {
+              newTarget {
+                name
+              }
+            }
+          }
+        """,
+        expected = Right(
+          json"""
+            {
+              "cloneTarget" : {
+                "newTarget" : {
+                  "name" : "V1647 Orionis"
+                }
+              }
+            }
+          """
+        )
+      )
+    yield ()
+  }
+
+}

